aqt/metadata.py

```diff
@@ -0,0 +1,657 @@
+            self.head = head
+            self.rows = rows
+            self.max_width = max_width
+
+        def pretty_print(self) -> str:
+            table = Texttable(max_width=self.max_width)
+            table.set_deco(Texttable.HEADER)
+            table.header(self.head)
+            table.add_rows(self.rows, header=False)
+            return table.draw()
+
+    def __init__(
+        self,
+        archive_id: ArchiveId,
+        *,
+        filter_minor: Optional[int] = None,
+        is_latest_version: bool = False,
+        modules_ver: Optional[str] = None,
+        extensions_ver: Optional[str] = None,
+        architectures_ver: Optional[str] = None,
+        tool_name: Optional[str] = None,
+        tool_long_listing: Optional[str] = None,
+    ):
+        """
+        Construct ListCommand.
+
+        :param filter_minor:        When set, the ListCommand will filter out all versions of
+                                    Qt that don't match this minor version.
+        :param is_latest_version:   When True, the ListCommand will find all versions of Qt
+                                    matching filters, and only print the most recent version
+        :param modules_ver:         Version of Qt for which to list modules
+        :param extensions_ver:      Version of Qt for which to list extensions
+        :param architectures_ver:   Version of Qt for which to list architectures
+        """
+        self.logger = getLogger("aqt.archives")
+        self.archive_id = archive_id
+        self.filter_minor = filter_minor
+
+        if archive_id.is_tools():
+            if tool_name:
+                self.request_type = "tool variant names"
+                self._action = lambda: self.fetch_tool_modules(tool_name)
+            elif tool_long_listing:
+                self.request_type = "tool long listing"
+                self._action = lambda: self.fetch_tool_long_listing(tool_long_listing)
+            else:
+                self.request_type = "tools"
+                self._action = self.fetch_tools
+        elif is_latest_version:
+            self.request_type = "latest version"
+            self._action = lambda: Versions(self.fetch_latest_version())
+        elif modules_ver:
+            self.request_type = "modules"
+            self._action = lambda: self.fetch_modules(self._to_version(modules_ver))
+        elif extensions_ver:
+            self.request_type = "extensions"
+            self._action = lambda: self.fetch_extensions(
+                self._to_version(extensions_ver)
+            )
+        elif architectures_ver:
+            self.request_type = "architectures"
+            self._action = lambda: self.fetch_arches(
+                self._to_version(architectures_ver)
+            )
+        else:
+            self.request_type = "versions"
+            self._action = self.fetch_versions
+
+    def action(self) -> Union[List[str], Versions]:
```

Comment:
This should probably have the type hint `-> Union[List[str], Versions, Table]`, because `Table` is what `action` returns in the case of `aqt list tools <host> <target> --tool-long <toolname>`. I forgot to add the Table type to the type hint for #304, and I really should have.